lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -789,73 +789,44 @@ static bool checkRequirementsSatisfied(
-        secondType = origSecondType;
-      // If the second type is a dependent member type, try to resolve it
-      // using the first type as the base type.
-      // TODO: Add checks verifying that first type can be the base type.
-      if (auto depMemType = secondType->getAs<DependentMemberType>()) {
-        if (!firstType->hasTypeParameter())
-          if (auto substType = depMemType->substBaseType(
-                  firstType, LookUpConformanceInModule(swiftModule)))
-            secondType = substType->getCanonicalType();
-      }
+    case RequirementKind::SameType:
+      // If the first type does not equal the second type, then record the
+      // unsatisfied requirement.
       if (!firstType->isEqual(secondType))
         unsatisfiedRequirements.push_back(req);
       continue;
-    }
     // Check conformance requirements.
     case RequirementKind::Conformance: {
       auto protocolType = req.getSecondType()->castTo<ProtocolType>();
       auto protocol = protocolType->getDecl();
       assert(protocol && "Expected protocol in generic signature requirement");
-      // Try substituting requirement type using original substutition map.
-      // If the result type is known to conform to protocol in the current
-      // module, continue.
-      // This handles cases where the VJP caller is non-generic (specialized
-      // with concrete types) but the associated derivative callee is generic.
-      if (auto origFirstType = firstType.subst(substMap)) {
-        if (!origFirstType->hasError() &&
-            swiftModule->lookupConformance(origFirstType, protocol)) {
-          continue;
-        }
-      }
-      Optional<ProtocolConformanceRef> conformance;
-      // FIXME terrible hack: If LHS is a dependent member, try to resolve it
-      // using the `Self` type, assuming the first replacement type is `Self`.
-      if (auto depMemType = firstType->getAs<DependentMemberType>()) {
-        if (original->hasSelfParam() &&
-            !substMap.getReplacementTypes().empty()) {
-          if (auto substType = depMemType->substBaseType(
-                  substMap.getReplacementTypes().front(),
-                  LookUpConformanceInModule(swiftModule))) {
-            firstType = substType;
-          }
-        }
-        conformance = swiftModule->conformsToProtocol(firstType, protocol);
-      }
-      // Otherwise, try to look up conformance in substitution maps.
-      else {
-        conformance = substMap.lookupConformance(
-            firstType->getCanonicalType(), protocol);
-      }
-        
-      if (!conformance)
+      // If the first type does not conform to the second type in the current
+      // module, then record the unsatisfied requirement.
+      if (!swiftModule->lookupConformance(firstType, protocol))
         unsatisfiedRequirements.push_back(req);
       continue;
     }
     // Ignore other requirements (superclass and layout).
+    // Layout requirements are rejected during type-checking.
```

Comment:
Is there no value in allowing layout requirements here? The only kind that is implemented so far is conformance to AnyObject, which seems useful.

Reply:
Is it the case that `T : AnyObject` is a layout requirement, not a conformance requirement?

It's not clear that supporting layout requirements is desirable for differentiation, but we're happy to revisit it after differentiation has better support for classes, and when differentiation is pitched via Swift Evolution.

Reply:
Correct. `T : AnyObject` is a layout requirement. AnyObject is not a protocol.